# Define the agents

In [35]:
import random
import math
from numpy.core.shape_base import atleast_2d
import numpy as np
import pandas as pd
import matplotlib.ticker as ticker
import matplotlib.pyplot as plt


import matplotlib.pyplot as plt
from matplotlib.patches import FancyBboxPatch



In [36]:
# Import functions in the function file

%run trade_functions.ipynb


## define encounter matrix

In [37]:
economy_size=1000

In [38]:
import numpy as np

# Create a 100x100 matrix filled with 1s
matrix = np.ones((economy_size, economy_size))

# And  Set the diagonal elements to 0
np.fill_diagonal(matrix, 0)


## Define Economy A

In [39]:
agent_a = create_agent()
agent_a['id_list'] = [i for i in range(economy_size)]
agent_a['parameter']=[[3,3,3] for _ in range(economy_size)]
agent_a['money']=[2 for _ in range(economy_size)]
agent_a['gooda']=[1 for _ in range(economy_size)]
agent_a['goodb']=[2 for _ in range(economy_size)]
agent_a['intra_connections']=matrix
agent_a['inter_connections']=matrix

# Define Economy B

In [40]:
agent_b = create_agent()
agent_b['id_list'] = [i for i in range(economy_size)]
agent_b['parameter']=[[3,3,3] for _ in range(economy_size)]
agent_b['money']=[2 for _ in range(economy_size)]
agent_b['gooda']=[1 for _ in range(economy_size)]
agent_b['goodb']=[2 for _ in range(economy_size)]
agent_b['intra_connections']=matrix
agent_b['inter_connections']=matrix


In [41]:
agent_c = create_agent()
agent_c['id_list'] = [i for i in range(economy_size)]
agent_c['parameter']=[[2,2,2] for _ in range(economy_size)]
agent_c['money']=[2 for _ in range(economy_size)]
agent_c['gooda']=[1 for _ in range(economy_size)]
agent_c['goodb']=[2 for _ in range(economy_size)]

agent_c['constant']=[1 for _ in range(economy_size)]
agent_c['intra_connections']=matrix
agent_c['inter_connections']=matrix

#Defne  inhomogeneous economy

In [42]:
def utility_func_cd(ga,gb,m,parameter):
  a=m**(parameter[0]-1)

  b=ga**(parameter[1]-1)

  c=gb**(parameter[2]-1)

  utility=a*b*c
  utility=float(utility)
  return utility

In [43]:
def utility_func_s(ga,gb,m,parameter):
  a=m**(parameter[0]-1)

  b=ga**(parameter[1]-1)

  c=gb**(parameter[2]-1)

  utility=a*(b+c)
  utility=float(utility)
  return utility

In [44]:
def utility_func_c(ga,gb,m,parameter):
  a=m**(parameter[0]-1)

  

  

  utility=a*(min(ga,gb)**(parameter[1]-1))
  utility=float(utility)
  return utility

## trade function

In [45]:
import random

def encounter1(matrix):
    # Pick a random row
    row = random.choice(range(len(matrix)))

    # Pick a random column different from the row
    possible_cols = [i for i in range(len(matrix[0])) if i != row]
    col = random.choice(possible_cols)

    return row, col







# metropolis

In [30]:
def onsager(agent1,agent2,length,size,t,f1,f2,flow):
  
  mla=agent1['money'].copy()
  gala=agent1['gooda'].copy()
  gbla=agent1['goodb'].copy()
  parametera=agent1['parameter'].copy()
    
  mlb=agent2['money'].copy()
  galb=agent2['gooda'].copy()
  gblb=agent2['goodb'].copy()
  parameterb=agent2['parameter'].copy()
  sum_etaa = sum(sublist[2] for sublist in parametera)
  sum_etab= sum(sublist[2] for sublist in parameterb)

 

  
  
  ga1_history=[]
  gb1_history=[]
  m1_history=[] 
  ga2_history=[]
  gb2_history=[]
  m2_history=[] 
  for i in range(length[0]):
    
    gala,gbla,mla=intra_trade(gala,gbla,mla,parametera,matrix,1,f1,f1)
    
    
    galb,gblb,mlb=intra_trade(galb,gblb,mlb,parameterb,matrix,1,f1,f2)
    

    
  
 
      
    
    
  
    


    
  for j in range(length[1]):
    
    gala,gbla,mla=intra_trade(gala,gbla,mla,parametera,matrix,1,f1,f2)
    gala,gbla,mla=direct_trade(gala,gbla,mla, 1, flow)
  
    
    galb,gblb,mlb=intra_trade(galb,gblb,mlb,parameterb,matrix,1,f1,f2)
    galb,gblb,mlb=direct_trade(galb,gblb,mlb, -1, flow)
  
    
    gala,gbla,mla,galb,gblb,mlb=inter_trade(gala,gbla,mla,galb,gblb,mlb,parametera,parameterb,matrix,size,t,f1,f2)

      
    ga1_history.append(sum(gala))
    gb1_history.append(sum(gbla))
    m1_history.append(sum(mla))
    ga2_history.append(sum(galb))
    gb2_history.append(sum(gblb))
    m2_history.append(sum(mlb))
   

  return ga1_history,gb1_history,m1_history,ga2_history,gb2_history,m2_history





In [31]:
def tester(agent1,agent2,gm,length,size,f1,f2):
  
  
  mla=[gm[2]]*economy_size
  gala=[gm[0]]*economy_size
  gbla=[gm[1]]*economy_size
  
  parametera=agent2['parameter'].copy()
    
  mlb=agent1['money'].copy()
  galb=agent1['gooda'].copy()
  gblb=agent1['goodb'].copy()
  constant=agent1['constant'].copy()
  parameterb=agent1['parameter'].copy()
  sum_eta=np.sum(agent1['parameter'][2])
  sum_alpha1=np.sum(agent1['parameter'][0])
  sum_alpha2=np.sum(agent1['parameter'][1])
 

  

  ga2_history=[]
  gb2_history=[]
  m2_history=[] 

  for i in range(length[0]):
    
    gala,gbla,mla=intra_trade(gala,gbla,mla,parametera,matrix,1,f1,f2)
        
      
        
    galb,gblb,mlb=intra_trade(galb,gblb,mlb,parameterb,matrix,1,f1,f2)
    
  
    
    gala,gbla,mla,galb,gblb,mlb=test_trade(gala,gbla,mla,galb,gblb,mlb,parametera,parameterb,matrix,1,[1,0,0],f1,f2)
    gala,gbla,mla,galb,gblb,mlb=test_trade(gala,gbla,mla,galb,gblb,mlb,parametera,parameterb,matrix,1,[0,1,0],f1,f2)
    gala,gbla,mla,galb,gblb,mlb=test_trade(gala,gbla,mla,galb,gblb,mlb,parametera,parameterb,matrix,1,[0,0,1],f1,f2)
    

    
  
 
      
    
    
  
    

  for k in range (8):
      size=size*0.7
    
      for j in range(length[1]):
        
        gala,gbla,mla=intra_trade(gala,gbla,mla,parametera,matrix,1,f1,f2)
        
      
        
        galb,gblb,mlb=intra_trade(galb,gblb,mlb,parameterb,matrix,1,f1,f2)
        
      
        
        gala,gbla,mla,galb,gblb,mlb=test_trade(gala,gbla,mla,galb,gblb,mlb,parametera,parameterb,matrix,size,[1,0,0],f1,f2)
        gala,gbla,mla,galb,gblb,mlb=test_trade(gala,gbla,mla,galb,gblb,mlb,parametera,parameterb,matrix,size,[0,1,0],f1,f2)
        gala,gbla,mla,galb,gblb,mlb=test_trade(gala,gbla,mla,galb,gblb,mlb,parametera,parameterb,matrix,size,[0,0,1],f1,f2)
    
        
       
        ga2_history.append(sum(galb))
        gb2_history.append(sum(gblb))
        m2_history.append(sum(mlb))
        

  return ga2_history,gb2_history,m2_history

In [32]:
def get_value(lst,sum_p,lenx):
    sub_list=lst[-lenx:]
    value_list=[ sum_p/x for x in sub_list]
    return np.mean(value_list)
    

In [33]:
def get_vgap(length1,length2,length3,range1,f1,flow):
    mga1,mgb1,mm1,mga2,mgb2,mm2=[],[],[],[],[],[]
    for i in range (length3):
        ga1,gb1,m1,ga2,gb2,m2=onsager(agent_a,agent_b,length1,1,[1,1,1],f1,f1,flow)
        mga1.append(np.mean(ga1[-range1:])/economy_size)
        mga1.append(np.mean(ga1[-range1:])/economy_size)
        mgb1.append(np.mean(gb1[-range1:])/economy_size)
        mm1.append(np.mean(m1[-range1:])/economy_size)
        mga2.append(np.mean(ga2[-range1:])/economy_size)
        mgb2.append(np.mean(gb2[-range1:])/economy_size)
        mm2.append(np.mean(m2[-range1:])/economy_size)
    
    gm1=[np.mean(mga1),np.mean(mgb1),np.mean(mm1)]
    gm2=[np.mean(mga2),np.mean(mgb2),np.mean(mm2)]
    print(np.mean(mga1),np.mean(mgb1),np.mean(mm1),np.mean(mga2),np.mean(mgb2),np.mean(mm2))
    mvga,mvgb,mvm,mvga2,mvgb2,mvm2=[],[],[],[],[],[]
    for j in range(length3):
        vga,vgb,vm=tester(agent_c,agent_a,gm1,length2,1,f1,utility_func_cd)
        vga2,vgb2,vm2=tester(agent_c,agent_b,gm2,length2,1,f1,utility_func_cd)
        mvga.append(get_value(vga,2000,10000))
        mvgb.append(get_value(vgb,2000,10000))
        mvm.append(get_value(vm,2000,10000))
        print(mvga,mvgb,mvm)
        mvga2.append(get_value(vga2,2000,10000))
        mvgb2.append(get_value(vgb2,2000,10000))
        mvm2.append(get_value(vm2,2000,10000))
        print(mvga2,mvgb2,mvm2)
    print('Gap',np.mean(mvga2)-np.mean(mvga),np.mean(mvgb2)-np.mean(mvgb),np.mean(mvm2)-np.mean(mvm))
    return np.mean(mvga2)-np.mean(mvga),np.mean(mvgb2)-np.mean(mvgb),np.mean(mvm2)-np.mean(mvm)
        

## Substitute economy

In [34]:
g12,g22,g32=get_vgap([10000,90000],[20000,20000],5,90000,utility_func_s,[0.2,0,0])

1.2011364641179016 1.9971199292512352 2.0045816984711964 0.7988635358820991 2.002880070748765 1.9954183015288032
[1.3617067036666568] [1.0897243613769863] [1.5155262639824005]
[1.8199282244093982] [1.1976472275605927] [1.4949783690604752]
[1.3617067036666568, 1.3624585898033654] [1.0897243613769863, 1.0916819299420548] [1.5155262639824005, 1.4995612461581007]
[1.8199282244093982, 1.8364628467912247] [1.1976472275605927, 1.194224512098863] [1.4949783690604752, 1.4986633521833337]
[1.3617067036666568, 1.3624585898033654, 1.3632522746685334] [1.0897243613769863, 1.0916819299420548, 1.0917458730945717] [1.5155262639824005, 1.4995612461581007, 1.5007286819132881]
[1.8199282244093982, 1.8364628467912247, 1.8216068543411905] [1.1976472275605927, 1.194224512098863, 1.1905637813022227] [1.4949783690604752, 1.4986633521833337, 1.4921551022017554]
[1.3617067036666568, 1.3624585898033654, 1.3632522746685334, 1.373030555434043] [1.0897243613769863, 1.0916819299420548, 1.0917458730945717, 1.08113191

In [47]:
g12,g22,g32=get_vgap([10000,90000],[20000,20000],5,90000,utility_func_s,[0,0.2,0])

0.9982349142684319 2.201922959598366 1.9965927883517285 1.0017650857315683 1.798077040401635 2.003407211648272
[1.5204861484033443] [1.0579320313652527] [1.5055799994796173]
[1.590433118567283] [1.2343367894981454] [1.5067638382530784]
[1.5204861484033443, 1.4972612096695253] [1.0579320313652527, 1.056820731364928] [1.5055799994796173, 1.4835244349413579]
[1.590433118567283, 1.5863616558292186] [1.2343367894981454, 1.2415102204754753] [1.5067638382530784, 1.4873702643711528]
[1.5204861484033443, 1.4972612096695253, 1.4972925779115553] [1.0579320313652527, 1.056820731364928, 1.0571484847953492] [1.5055799994796173, 1.4835244349413579, 1.507111309180378]
[1.590433118567283, 1.5863616558292186, 1.6167897244106268] [1.2343367894981454, 1.2415102204754753, 1.2354366811788358] [1.5067638382530784, 1.4873702643711528, 1.5004498995643423]
[1.5204861484033443, 1.4972612096695253, 1.4972925779115553, 1.5103866627524742] [1.0579320313652527, 1.056820731364928, 1.0571484847953492, 1.06476252090570

In [48]:
g13,g23,g33=get_vgap([10000,90000],[20000,20000],5,90000,utility_func_s,[0,0,0.2])

0.9990078771522626 2.000418415801845 2.197767486593158 1.0009921228477374 1.999581584198155 1.8022325134068435
[1.542972730069829] [1.1371065517915633] [1.3641666262790815]
[1.5459268291743236] [1.1460380984518501] [1.664759902503423]
[1.542972730069829, 1.553479012380897] [1.1371065517915633, 1.1300225867485152] [1.3641666262790815, 1.3661755733192797]
[1.5459268291743236, 1.5559199743405563] [1.1460380984518501, 1.1342516992858414] [1.664759902503423, 1.6702736035054158]
[1.542972730069829, 1.553479012380897, 1.5535088657141551] [1.1371065517915633, 1.1300225867485152, 1.130301768542614] [1.3641666262790815, 1.3661755733192797, 1.3663927509007645]
[1.5459268291743236, 1.5559199743405563, 1.5560088051021228] [1.1460380984518501, 1.1342516992858414, 1.1299769446175352] [1.664759902503423, 1.6702736035054158, 1.6656546188890418]
[1.542972730069829, 1.553479012380897, 1.5535088657141551, 1.5484847793091436] [1.1371065517915633, 1.1300225867485152, 1.130301768542614, 1.1381668923162598] [

## Complement Economy

In [49]:
g11c,g21c,g31c=get_vgap([10000,90000],[20000,20000],5,90000,utility_func_c,[0.2,0,0])

1.1962277497294411 1.9975695164625784 1.9993981478748801 0.8037722502705593 2.0024304835374216 2.00060185212512
[1.9803693355619576] [0.7671497840855633] [1.4899441706969239]
[3.207224198951686] [0.6687082019054024] [1.5097824474030952]
[1.9803693355619576, 1.987525547482522] [0.7671497840855633, 0.7641719262659129] [1.4899441706969239, 1.4839999836191176]
[3.207224198951686, 3.223224137419082] [0.6687082019054024, 0.6613329108294358] [1.5097824474030952, 1.5024900800918017]
[1.9803693355619576, 1.987525547482522, 2.000603128660723] [0.7671497840855633, 0.7641719262659129, 0.7675014756379195] [1.4899441706969239, 1.4839999836191176, 1.4901984185465134]
[3.207224198951686, 3.223224137419082, 3.2114320366769205] [0.6687082019054024, 0.6613329108294358, 0.6640634514314115] [1.5097824474030952, 1.5024900800918017, 1.4940630480752828]
[1.9803693355619576, 1.987525547482522, 2.000603128660723, 1.9941009130429237] [0.7671497840855633, 0.7641719262659129, 0.7675014756379195, 0.761644071167716]

In [50]:
g12c,g22c,g32c=get_vgap([10000,90000],[20000,20000],5,90000,utility_func_c,[0,0.2,0])


1.0000383621609614 2.1960857134792677 2.001140985119064 0.9999616378390386 1.8039142865207336 1.9988590148809362
[2.5432826844173766] [0.6308957850171164] [1.508005953757474]
[2.4239286782279295] [0.824898779055041] [1.499116497640796]
[2.5432826844173766, 2.5513781496009753] [0.6308957850171164, 0.6259759812725054] [1.508005953757474, 1.5038303693494517]
[2.4239286782279295, 2.4339748472960934] [0.824898779055041, 0.8356069299000582] [1.499116497640796, 1.4960890688584496]
[2.5432826844173766, 2.5513781496009753, 2.535769331055475] [0.6308957850171164, 0.6259759812725054, 0.619239444225975] [1.508005953757474, 1.5038303693494517, 1.5007987300605563]
[2.4239286782279295, 2.4339748472960934, 2.424586505565351] [0.824898779055041, 0.8356069299000582, 0.8215777151116221] [1.499116497640796, 1.4960890688584496, 1.5118250484729645]
[2.5432826844173766, 2.5513781496009753, 2.535769331055475, 2.5629574464626503] [0.6308957850171164, 0.6259759812725054, 0.619239444225975, 0.6246184652387474] [


Gap 1.2379799731607277 -0.09901665282173466 0.009414949860635069
Gap -0.11015155599964821 0.19847037605772722 0.00798941294173705
Gap -0.01663484211594346 -0.002696506245851049 0.2992208639674707

In [52]:
g13c,g23c,g33c=get_vgap([10000,90000],[20000,20000],5,90000,utility_func_c,[0,0,0.2])

0.9989949548485642 1.99911589255363 2.1960890551202787 1.0010050451514356 2.00088410744637 1.803910944879723
[2.492315128939205] [0.7157710965702] [1.370506323677323]
[2.480094470388574] [0.7103527488810906] [1.6481110178081053]
[2.492315128939205, 2.4852091637156453] [0.7157710965702, 0.7212951270799112] [1.370506323677323, 1.3519044700812355]
[2.480094470388574, 2.492072665094125] [0.7103527488810906, 0.7109911221552404] [1.6481110178081053, 1.670078297742214]
[2.492315128939205, 2.4852091637156453, 2.4939136338189862] [0.7157710965702, 0.7212951270799112, 0.7071704992120202] [1.370506323677323, 1.3519044700812355, 1.3718018717878353]
[2.480094470388574, 2.492072665094125, 2.4795069743828977] [0.7103527488810906, 0.7109911221552404, 0.7137847290731577] [1.6481110178081053, 1.670078297742214, 1.6644201913842809]
[2.492315128939205, 2.4852091637156453, 2.4939136338189862, 2.5193089971532956] [0.7157710965702, 0.7212951270799112, 0.7071704992120202, 0.7153060930870738] [1.37050632367732

# Calculate L matrix

In [ ]:
Gap 0.45738226301711915 0.10492674938835389 -0.012332599563730007
Gap 0.08993564849956126 0.17815757916991815 -0.0010839639875976914
Gap 0.0034277867724332722 0.00225657814324709 0.29825261255819235

In [ ]:
Gap 1.2379799731607277 -0.09901665282173466 0.009414949860635069
Gap -0.11015155599964821 0.19847037605772722 0.00798941294173705
Gap -0.01663484211594346 -0.002696506245851049 0.2992208639674707

## Substitute

In [55]:
print('Lmm=fm/dbeta=', 0.2/g33)

Lmm=fm/dbeta= 0.670572499883728


In [53]:
0.2=a*dv1+b*dv2
0=c*dv1+d*dv2

0=a*dv1+b*dv2
0.2=c*dv1+d*dv2

67.05724998837279

In [ ]:
0.2=a*0.457+b*0.105
0=c*0.457+d*0.105
0=a*0.090+b*0.178
0.2=c*0.090+d*0.178

In [56]:
import numpy as np

# Coefficient matrix
A = np.array([
    [0.457, 0.105, 0, 0],
    [0, 0, 0.457, 0.105],
    [0.090, 0.178, 0, 0],
    [0, 0, 0.090, 0.178]
])

# Constant matrix
B = np.array([0.2, 0, 0, 0.2])

# Solving the system of equations
solution = np.linalg.solve(A, B)

# Output the solution
a, b, c, d = solution
print(f"a = {a:.4f}")
print(f"b = {b:.4f}")
print(f"c = {c:.4f}")
print(f"d = {d:.4f}")


a = 0.4952
b = -0.2504
c = -0.2921
d = 1.2713


## Complement

In [57]:
import numpy as np

# Coefficient matrix
A = np.array([
    [g11c, g21c, 0, 0],
    [0, 0, g11c, g21c],
    [g12c, g22c, 0, 0],
    [0, 0, g12c, g22c]
])

# Constant matrix
B = np.array([0.2, 0, 0, 0.2])

# Solving the system of equations
solution = np.linalg.solve(A, B)

# Output the solution
a, b, c, d = solution
print(f"a = {a:.4f}")
print(f"b = {b:.4f}")
print(f"c = {c:.4f}")
print(f"d = {d:.4f}")


a = 0.1691
b = 0.0938
c = 0.0843
d = 1.0545


In [58]:
print('Lmm=fm/dbeta=', 0.2/g33c)

Lmm=fm/dbeta= 0.6684025884697087


$$ 
\begin{bmatrix}
0.6705 & \begin{bmatrix} 0, 0 \end{bmatrix} \\
\begin{bmatrix} 0 \\ 0 \end{bmatrix} & \begin{bmatrix} 0.4952 & -0.2504 \\ -0.2921 & 1.2713 \end{bmatrix}
\end{bmatrix}
$$


$$ 
\begin{bmatrix}
0.6684 & \begin{bmatrix} 0, 0 \end{bmatrix} \\
\begin{bmatrix} 0 \\ 0 \end{bmatrix} & \begin{bmatrix} 0.1691 & 0.0938 \\ 0.0843 & 1.0545 \end{bmatrix}
\end{bmatrix}
$$